[MRG] Use hashtable (python sets/dicts) for object dtype data in Encoders

jorisvandenbossche opened the pull request:
#### Reference Issues/PRs
Fixes #7432, see also https://github.com/scikit-learn/scikit-learn/pull/9151

#### What does this implement/fix?

This implements the encoding in the LabelEncoder in the different ways (the original numpy/searchsorted one, one based on sets/dict lookup (much faster for object dtype) and one using pandas), and dispatching to one of those methods depending on the dtype / depending on whether pandas is available.

This is WIP, mainly to get feedback on the idea. I am not happy yet with the implementation. Also for `fit_transform` there are shortcuts available for the numpy and pandas way, which I didn't add yet.

jorisvandenbossche committed:
Optionally use hashtable or pandas for LabelEncoder

jnothman commented:
This pull request introduces 1 alert - [view on lgtm.com](https://lgtm.com/projects/g/scikit-learn/scikit-learn/rev/pr-d245d67d859fee2a20a8aabba8e8bcbd2f3980f3)

**new alerts:**

* 1 for Wrong number of arguments in a call

---

*Comment posted by [lgtm.com](https://lgtm.com)*

jnothman commented:
Throw us a quick benchmark for strings, ints, few categories, many categories? Also, a comment on the unsorted case (not that I'm sure we want that in LabelEncoder, really...)?

jorisvandenbossche commented:
There are quite extensive benchmarks in the notebook I linked to in the issue: http://nbviewer.jupyter.org/gist/jorisvandenbossche/f399d43499785534f018a4f0c16d24dd

Two summary plots:

Factorizing step (`fit`) (time in seconds on 1 million elements), 'LabelEncoder' and 'numpy' are basically the same, and you can ignore the 'dict' one:

![index](https://user-images.githubusercontent.com/1020496/33267409-a91c584c-d379-11e7-90f3-eff34b92753e.png)

Encoding step (`transform`) (time in seconds on 1 million elements), sorry for the colors but int and object are switched in comparison to the plot above

![index2](https://user-images.githubusercontent.com/1020496/33267411-ab46d278-d379-11e7-87d4-dc5b005742f9.png)

jorisvandenbossche commented:
I updated the figures above a bit as they turned out to be not fully correct for the transform step (the LabelEncoder does more than just encoding with numpy (which I wanted to measure), it also checks if there are unseen categories. Which is something I did not time for the other methods, so added a separate numpy one that also does not do that. 
Will probably need to add this checking for unseen labels to the benchmarks as well.

The conclusion can also that the numpy method is quite OK, except for object dtype. So just special casing object dtype with a set/dict based one might be enough (and this one can then be used in the CategoricalEncoder when the categories are not sorted).

amueller commented:
why do we need the table if we don't have the unsorted case?

amueller commented:
If we only do numpy & python, we would use numpy for strings? So this wouldn't really impact CategoricalEncoder in most cases, right?

jorisvandenbossche commented:
> why do we need the table if we don't have the unsorted case?

Because it is much faster for object dtype + we want to support unsorted case in CategoricalEncoder

amueller commented:
Does unsorted mean they are in a different order or they are not sortable?

jorisvandenbossche commented:
> If we only do numpy & python, we would use numpy for strings? So this wouldn't really impact CategoricalEncoder in most cases, right?

Yes, I suppose most people with strings will actually have object dtype instead of numpy string dtype (= all people using pandas to store strings)

jorisvandenbossche commented:
For the CategoricalEncoder, I think ideally we want to allow user-specified categories that do not need to be sorted (now the sortedness is imposed and an error is raised if it is not the case). I think normally they will be sortable though, but so it is about being in a different order.

jnothman commented:
> The conclusion can also that the numpy method is quite OK, except for object dtype. So just special casing object dtype with a set/dict based one might be enough (and this one can then be used in the CategoricalEncoder when the categories are not sorted).

On the other hand, users providing object arrays with strings will very often be doing that through Pandas. But yes, I agree with you that we need the Python implementation anyway.

My only remaining hesitation about not using pandas is that this encoding is done all over the place (i.e. in every classifier), even repeatedly encoding the same data, and we may be wasting seconds of predict time by not adopting Pandas.

Btw, you may be catching a Pandas fast path for your int tests by making the set of classes 0..n-1 rather than something gappy or starting above/below 0. It means that the transform is a no-op.

jorisvandenbossche committed:
clean-up

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into labelencoder-sets

glemaitre commented on sklearn/preprocessing/tests/test_label.py:
```diff
@@ -536,3 +541,33 @@ def test_inverse_binarize_multiclass():
+def test_factorize_encode_utils(engine, values, expected):
+    # test that all different encoders are equivalent
+
+    if engine == 'numpy':
```
you can parametrize this part as well and mark one of the solution as skipif with pytest, isn't it?

glemaitre commented on sklearn/preprocessing/label.py:
```diff
@@ -94,24 +177,24 @@ def fit(self, y):
-        y = column_or_1d(y, warn=True)
-        self.classes_, y = np.unique(y, return_inverse=True)
-        return y
+    # def fit_transform(self, y):
```
clean up

glemaitre commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,89 @@
+    return _PANDAS_INSTALLED
+
+
+# def _encode_numpy(values, uniques=None, encode=True):
```
clean up

glemaitre reviewed (changes requested):
Then we need a what's new and the factorize seems to not work with read-only memmap.

jorisvandenbossche committed:
remove pandas for now; simplify; document

jorisvandenbossche committed:
move detection unseen labels into encode function

jnothman reviewed (commented):
I think I'd like to see all LabelEncoder tests run on both number and object dtypes...

jorisvandenbossche committed:
use new _encode instead of LabelEncoder in CategoricalEncoder

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,76 @@
+    Returns
+    -------
+    uniques
+        If decode=False
```
decode isn't a thing.

jnothman reviewed (commented):
Is the intention of putting `_encode` directly into `CategoricalEncoder`, rather than `LabelEncoder`, in order to allow for unordered mappings?

jorisvandenbossche committed:
parametrize more tests

jorisvandenbossche commented:
I think I parametrized now all LabelEncoder tests that are useful to parametrize for both int and object.

jorisvandenbossche commented:
> Is the intention of putting _encode directly into CategoricalEncoder, rather than LabelEncoder, in order to allow for unordered mappings?

Yes, see the one but last commit I added. 
For now I just replaced LabelEncoder with `_encode`, but the idea is indeed that in that way I can relax the restriction on sorted categories (at least for object dtype). Working on that next (but first need to fix handling of unknown categories in CategoricalEncoder as well)

jnothman commented:
let's merge the CategoricalEncoder rewrite first then. can you go fix that
0.19dev to 0.20dev? I'm on the move.

jorisvandenbossche committed:
also properly handle unknown categories in CategoricalEncoder

jorisvandenbossche committed:
reuse check function for LabelEncoder as well

jorisvandenbossche commented:
I added an extra function to check unknown values with numpy/python for non-object/object dtypes. I will probably have to parametrize some CategoricalEncoder tests for this to ensure both paths are properly tested, but will only do that when the CategoricalEncoder rewrite is merged (will update that now).

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into labelencoder-sets

jorisvandenbossche committed:
allow unsorted categories passed by user for object dtype

jorisvandenbossche commented:
cc @ogrisel This should be ready enough now to review (I still need to check if I need to parametrize more OneHotEncoder tests to ensure they both use numerical and object dtype data, but will only get to that tomorrow)

jorisvandenbossche committed:
parametrize some OneHotEncoder tests for object/int dtypes

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -351,7 +351,12 @@ def check_categorical_onehot(X):
+    np.array([['b', 'A', 'cat'], ['a', 'B', 'cat']], dtype=object)
+    ], ids=['mixed', 'numeric', 'object'])
+def test_one_hot_encoder(X):
     X = [['abc', 1, 55], ['def', 2, 55]]
```
Uhm. I think that this should be removed.

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -404,36 +409,50 @@ def test_one_hot_encoder_inverse():
+    # when specifying categories manually, unknown categories should already
+    # raise when fitting
+    enc = OneHotEncoder(categories=cats)
+    assert_raises(ValueError, enc.fit, X2)
```
`pytest.raises`

glemaitre replied:
?

glemaitre commented on sklearn/preprocessing/tests/test_encoders.py:
```diff
@@ -445,14 +464,23 @@ def test_one_hot_encoder_specified_categories():
+    X = np.array([[1, 2]]).T
+    enc = OneHotEncoder(categories=[[2, 1, 3]])
+    msg = re.escape('Unsorted categories are not supported')
+    assert_raises_regex(ValueError, msg, enc.fit_transform, X)
```
`pytest.raises` we could also match the error message.

glemaitre commented on sklearn/preprocessing/tests/test_label.py:
```diff
@@ -169,8 +172,33 @@ def test_label_binarizer_errors():
+    assert_array_equal(ret, [1, 0, 2, 0, 2])
+
+    msg = "unseen labels"
+    assert_raise_message(ValueError, msg, le.transform, unknown)
```
`pytest.raises`

glemaitre commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,126 @@
+        return _encode_numpy(values, uniques, encode)
+
+
+def _encode_check_unknown(values, uniques, return_mask=False):
```
I am not sure that we should start with `_` since that this is used in different file. We are not really consistent by checking the code base. @jnothman what do you recommend?

jnothman replied:
No problem having private internal utils

glemaitre reviewed (commented):
Couple of comments. Otherwise it looks good.

jorisvandenbossche commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,126 @@
+
+    Parameters
+    ----------
+    values : array
```
only numpy array (first thing I do is checking `.dtype`), but this is only used internally and thus we control it is always passed a numpy array

jorisvandenbossche committed:
feedback guillaume

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into labelencoder-sets

jorisvandenbossche commented:
Updated the PR with master; this should be ready for final review / merging.

jnothman commented:
Test failing.

jorisvandenbossche committed:
fixup merge master

jorisvandenbossche commented:
Whoops, sorry, rebasing error, should be fixed now.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+        encoded = np.searchsorted(uniques, values)
+        return uniques, encoded
+    else:
+        return uniques
```
Should "Unsorted categories are not supported for numerical categories" be validated and raised here?

jnothman replied:
Otherwise please note in the `_encode` docstring that this is not ensured, but is an assumption.

jorisvandenbossche replied:
It could be validated here (and it might be easier to follow), but in principle this will give (a little bit) more overhead as it is not needed to do that on each transform (encoding) if it is ensured in fit. 
So for now making this more clear in the docstring.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
 ]
 
 
+def _encode_numpy(values, uniques=None, encode=False):
```
Add a note that this should be accessed through `_encode`, where parameters are described.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+
+def _encode(values, uniques=None, encode=False):
+    """
+    Helper function to factorize (find uniques) and encode values.
```
our convention usually puts this summary on the previous line.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+    Uses pure python method for object dtype, and numpy method for
+    all other dtypes.
+    The numpy method has the limitation that the `uniques` need to
+    be sorted.
```
clarify that this is not validated, but assumed for all non-object input.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+    Returns
+    -------
+    uniques
+        If ``encode=False``.
```
note: Sorted if uniques parameter was None.

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+        diff = _encode_check_unknown(values, uniques)
+        if diff:
+            raise ValueError(
+                "y contains previously unseen labels: %s" % str(diff))
```
our style tends to prefer:
```py
            raise ValueError("y contains previously unseen labels: %s" %
                             str(diff))
```

jnothman commented on sklearn/preprocessing/label.py:
```diff
@@ -37,6 +37,125 @@
+            return diff
+    else:
+        unique_values = np.unique(values)
+        diff = list(np.setdiff1d(unique_values, uniques))
```
assume_unique=True

jnothman commented on sklearn/preprocessing/_encoders.py:
```diff
@@ -195,8 +194,8 @@ class OneHotEncoder(_BaseEncoder):
         - list : ``categories[i]`` holds the categories expected in the ith
-          column. The passed categories must be sorted and should not mix
-          strings and numeric values.
+          column. The passed categories should not mix strings and numeric
```
clarify that this is "within a single feature".

jnothman reviewed (commented):
Nitpicks. I've not checked tests, sorry.

jnothman reviewed (commented):
Is there a test for ohe with specified, unsorted cats?

jorisvandenbossche commented:
> Is there a test for ohe with specified, unsorted cats?

Yes, there is. See `test_one_hot_encoder_unsorted_categories`

jorisvandenbossche committed:
feedback Joel

jnothman commented:
> Yes, there is. See test_one_hot_encoder_unsorted_categories

Ah of course. Sorry I missed that.

jorisvandenbossche commented:
Tests are passing now.

jnothman commented:
@glemaitre your review here has a cross

glemaitre commented:
Is it worth to make an entry in the what's new in the enhancement section, thought it does not change anything for the user?

jorisvandenbossche commented:
I don't think a separate whatsnew entry is necessarily needed. I would see it as part of the enhancement to OneHotEncoder to support string data

ogrisel commented:
LGTM as well, merged!

jnothman commented:
cool!!

jorisvandenbossche commented:
Thanks!
